main: Validate config files by default

beautifulentropy opened the pull request:
- Make config validation run by default for all Boulder components with a registered validator.
- Refactor main to parse `boulder` flags directly instead of declaring them as subcommands.
- Remove the `validate` subcommand and update relevant docs.
- Fix configuration validation for issuer (file source) OCSP responder.

Fixes #6857
Fixes #6763

jsha commented on cmd/boulder/main.go:
```diff
@@ -60,57 +55,102 @@ func readAndValidateConfigFile(name, filename string) error {
+		err := readAndValidateConfigFile(core.Command(), configPath)
 		if err != nil {
-			fmt.Fprintf(os.Stderr, "Error validating configuration: %s\n", err)
+			fmt.Fprintf(os.Stderr, "Error validating config file %q: %v\n", configPath, err)
```
```suggestion
			fmt.Fprintf(os.Stderr, "Error validating config file %q: %s\n", configPath, err)
```

Generally better to use the %s output for errors.

jsha reviewed (changes requested):
Thanks for writing this!

This removes the validate subcommand entirely, but #6763 describes moving it. I thought the plan was that we would continue to have a `-validate` flag (which validates the config and then exits), but we would additionally run the validation during normal startup.

There's some manual handling of the new global flag `--skip-config-validation`. Is there a reason it wouldn't work to declare a `flag.Bool("skip-config-validation", ...` at the top level?

Also, this flag seems particularly risky. Now that Boulder is going to use this validator on all configs at startup, it seems likely we will rely on it within Boulder to ensure certain properties of operation. For instance we don't currently use config validation to ensure certificate lifetimes are within a BR-mandated range, but we might decide to do so in the future.

I could imagine a scenario where we added a new field, and the validation was overly strict so it couldn't actually be enabled. It would tempting in that case for SRE to override config validation in order to continue enabling the field, but I think our actual course should be to treat it as any other bug, and delay configuring the new field until a new Boulder release is made that fixes the validation.

beautifulentropy commented:
(summarizing our discussion from standup for posterity)

> This removes the validate subcommand entirely, but #6763 describes moving it. I thought the plan was that we would continue to have a `-validate` flag (which validates the config and then exits), but we would additionally run the validation during normal startup.

Based on feedback from SRE we've removed it because they don't plan to use it for ad-hoc validation.
 
> There's some manual handling of the new global flag `--skip-config-validation`. Is there a reason it wouldn't work to declare a `flag.Bool("skip-config-validation", ...` at the top level?

Yes, this is admittedly funky, however calling `flag.Parse()` before we get to the `notmain.main()` of a particular subcommand causes the wrong usage message to be displayed when an operator passes `--help/ -help`. We've got a similar problem here with `--config/ -config` as well. Each subcommand implements their own config, and some also don't or simply don't _require_ that a config be passed.

> Also, this flag seems particularly risky. Now that Boulder is going to use this validator on all configs at startup, it seems likely we will rely on it within Boulder to ensure certain properties of operation. For instance we don't currently use config validation to ensure certificate lifetimes are within a BR-mandated range, but we might decide to do so in the future.
> 
> I could imagine a scenario where we added a new field, and the validation was overly strict so it couldn't actually be enabled. It would tempting in that case for SRE to override config validation in order to continue enabling the field, but I think our actual course should be to treat it as any other bug, and delay configuring the new field until a new Boulder release is made that fixes the validation.

Thanks for your feedback; I think you're right. I've discussed this with SRE and they agreed to dropping the requirement.

aarongable commented on cmd/boulder/main.go:
```diff
@@ -60,57 +56,89 @@ func readAndValidateConfigFile(name, filename string) error {
+`,
+	core.Command())
+
+func main() {
```
The order of the three main stanzas in this function is:
1) Look up the subcommand if the executable name was "boulder"
2) Validate config (which 
3) Look up the subcommand if the executable name was not "boulder"

This feels weird to me. Also, I'm think it's broken (see comment below).

It seems like it should be
1) Figure out the name of the command that is being run
2) Look up the config struct, do config validation
3) Look up the function, execute it

aarongable commented on cmd/boulder/main.go:
```diff
@@ -60,57 +56,89 @@ func readAndValidateConfigFile(name, filename string) error {
+	configPath := getConfigPath()
+	if configPath != "" {
+		// Config flag passed.
+		err := readAndValidateConfigFile(core.Command(), configPath)
```
If `core.Command()` is "boulder" (i.e. we just executed the if clause above), isn't this going to fail to find a config struct to validate against?

beautifulentropy replied:
😅 yep, I used to assign `os.Args = os.Args[1:]` above this line but moved it recently and now it's definitely broken. I'll refactor this.

aarongable commented on docs/config-validation.md:
```diff
@@ -5,24 +5,27 @@ at https://github.com/letsencrypt/validator.
 
-For a complete list of `boulder` components which support config validation you
-can run:
+To skip validation, pass the `-skip-config-validation` flag:
```
This documentation should be removed now.

aarongable commented on cmd/boulder/main.go:
```diff
@@ -85,9 +85,11 @@ var boulderUsage = fmt.Sprintf(`Usage: %s <subcommand> [flags]
+	var command string
 	if core.Command() == "boulder" {
-		// The operator is running the boulder binary directly.
+		// Operator passed the boulder component as a subcommand.
+		command = os.Args[1]
```
I'd do this assignment on line 113, after the len/help/list checks. This var only gets used if those returns all get skipped, and it's kinda confusing to assing os.Args[1] to a variable, then reference os.Args[1] a bunch of times below rather than referencing the variable.

aarongable reviewed (commented):
LGTM modulo one change

jsha commented on test/issuer-ocsp-responder.json:
```diff
@@ -1,18 +1,38 @@
-   "stdoutlevel": 6,
-   "sysloglevel": 6
- }
+	"ocspResponder": {
```
This is the config for the issuer OCSP responder, which uses a file source, so it shouldn't require a redis stanza. Perhaps the validation for the OCSP responder config is wrong, and the redis stanza needs to be optional?
